### PR TITLE
add RegisterCheck server option for internal health checks

### PR DESCRIPTION
**Pull request by vtolstov**
Signed-off-by: Vasiliy Tolstov <v.tolstov@unistack.org>

### CODE REVIEW

**Comment by vtolstov** [comment]
i'm test it and it works fine for me (deregister forcible when error happened)

File: server/options.go [hunk 1]
```diff
@@ -163,6 +165,13 @@ func Metadata(md map[string]string) Option {
 	}
 }
 
+// RegisterCheck run func before registry service
```
Comment (asim) [inline server/options.go:172]: "RegisterCheck runs a check function before registering the service"

File: server/rpc_server.go [hunk 2]
```diff
@@ -518,8 +527,25 @@ func (s *rpcServer) Start() error {
...
+				// if register check func defined, use it
+				if fn := s.opts.RegisterCheck; fn != nil {
+					if err = fn(s.opts.Context); err != nil {
+						log.Logf("Server %s-%s register error: %s", config.Name, config.Id, err)
+						s.RLock()
+						registered := s.registered
+						s.RUnlock()
+						if registered {
+							// deregister self in case of error
+							if err := s.Deregister(); err != nil {
```
Comment (asim) [inline server/rpc_server.go]: You've not deregistered and not set s.registered to false. This code block is also a bit unclean in a huge if statement. Maybe there should be a default check function that always returns nil so we don't actually have to test the function exists.

Comment (vtolstov) [reply server/rpc_server.go]: i think that i fix this with new commit

**Comment by vtolstov** [comment]
i'm fix all, i think now it less spams in case of repeated errors in check func

File: server/options.go [hunk 3]
```diff
@@ -163,6 +169,13 @@ func Metadata(md map[string]string) Option {
...
```
Comment (asim) [inline server/options.go:172]: Can you make this comment match the above

Comment (vtolstov) [reply server/options.go:172]: yes, sorry. fixed

**Pull request merged**